add new APIs

showuon opened the pull request:
Added APIs:
 - ADD_TOPICS_TO_MIRROR
 - REMOVE_TOPICS_FROM_MIRROR
 - BUMP_LEADER_EPOCH

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

showuon committed:
add new APIs

showuon committed:
renmae the creatMirrorTopic API to attachMirrorTopic APi

showuon committed:
update attachMirrorTopic api

fvaleri commented on clients/src/main/java/org/apache/kafka/clients/admin/ForwardingAdmin.java:
```diff
@@ -280,6 +280,16 @@ public CreateMirrorResult createMirror(String mirrorName, Map<String, String> co
+    }
+
+    @Override
+    public AttachMirrorTopicResult attachMirrorTopic(Map<String, String> topicToMirrorName, AttachMirrorTopicOptions options) {
```
Should we call it `addTopicsToMirror`? Same for the API.

fvaleri commented on clients/src/main/java/org/apache/kafka/clients/admin/ForwardingAdmin.java:
```diff
@@ -280,6 +280,16 @@ public CreateMirrorResult createMirror(String mirrorName, Map<String, String> co
     }
 
+    @Override
+    public DeleteMirrorTopicResult deleteMirrorTopic(String clusterLinkName, Set<String> topics, DeleteMirrorTopicOptions options) {
```
Should we call it `removeTopicsFromMirror` to avoid any confusion? Same for the API.

fvaleri commented on clients/src/main/resources/common/message/DeleteMirrorTopicRequest.json:
```diff
@@ -0,0 +1,33 @@
+  "apiKey":96,
+  "type": "request",
+  "listeners": ["broker", "controller"],
+  "name": "DeleteMirrorTopicRequest",
```
```suggestion
  "name": "RemoveTopicsFromMirrorRequest",
```

fvaleri commented on clients/src/main/resources/common/message/DeleteMirrorTopicRequest.json:
```diff
@@ -0,0 +1,33 @@
+  "validVersions": "0",
+  "flexibleVersions": "0+",
+  "fields": [
+    { "name": "clusterLink", "type": "string", "versions": "7+", "ignorable": true,
```
```suggestion
    { "name": "mirrorName", "type": "string", "versions": "7+", "ignorable": true,
```

fvaleri commented on clients/src/main/resources/common/message/DeleteMirrorTopicRequest.json:
```diff
@@ -0,0 +1,33 @@
+    { "name": "Topics", "type": "[]TopicState", "versions": "0+", "about": "The name or topic ID of the topic.",
+      "fields": [
+        { "name": "TopicId", "type": "uuid", "versions": "0+", "about": "The unique topic ID."},
+        { "name": "Name", "type": "string", "versions": "0+", "mapKey": true, "entityType": "topicName",
```
```suggestion
        { "name": "TopicName", "type": "string", "versions": "0+", "mapKey": true, "entityType": "topicName",
```

fvaleri commented on clients/src/main/resources/common/message/DeleteMirrorTopicRequest.json:
```diff
@@ -0,0 +1,33 @@
+  "fields": [
+    { "name": "clusterLink", "type": "string", "versions": "7+", "ignorable": true,
+      "about": "If true, check that the topics can be created as specified, but don't create anything." },
+    { "name": "Topics", "type": "[]TopicState", "versions": "0+", "about": "The name or topic ID of the topic.",
```
```suggestion
    { "name": "Topics", "type": "[]TopicState", "versions": "0+", "about": "The topic state.",
```

fvaleri commented on clients/src/main/resources/common/message/DeleteMirrorTopicResponse.json:
```diff
@@ -0,0 +1,31 @@
+{
+  "apiKey":96,
+  "type": "response",
+  "name": "DeleteMirrorTopicResponse",
```
```suggestion
  "name": "RemoveTopicsFromMirrorResponse",
```

fvaleri commented on clients/src/main/resources/common/message/AttachMirrorTopicResponse.json:
```diff
@@ -0,0 +1,31 @@
+{
+  "apiKey":95,
+  "type": "response",
+  "name": "AttachMirrorTopicResponse",
```
```suggestion
  "name": "AddTopicsToMirrorResponse",
```

fvaleri commented on clients/src/main/resources/common/message/AttachMirrorTopicRequest.json:
```diff
@@ -0,0 +1,33 @@
+  "apiKey":95,
+  "type": "request",
+  "listeners": ["broker", "controller"],
+  "name": "AttachMirrorTopicRequest",
```
```suggestion
  "name": "AddTopicsToMirrorRequest",
```

fvaleri commented on clients/src/main/resources/common/message/AttachMirrorTopicRequest.json:
```diff
@@ -0,0 +1,33 @@
+    { "name": "Topics", "type": "[]TopicState", "versions": "0+", "about": "The name or topic ID of the topic.",
+      "fields": [
+        { "name": "TopicId", "type": "uuid", "versions": "0+", "about": "The unique topic ID."},
+        { "name": "Name", "type": "string", "versions": "0+", "mapKey": true, "entityType": "topicName",
```
```suggestion
        { "name": "TopicName", "type": "string", "versions": "0+", "mapKey": true, "entityType": "topicName",
```

fvaleri commented on clients/src/main/resources/common/message/BumpLeaderEpochRequest.json:
```diff
@@ -0,0 +1,34 @@
+  "validVersions": "0",
+  "flexibleVersions": "0+",
+  "fields": [
+    { "name": "Topics", "type": "[]TopicState", "versions": "0+", "about": "The name or topic ID of the topic.",
```
```suggestion
    { "name": "Topics", "type": "[]TopicState", "versions": "0+", "about": "The topic and partitions state.",
```

fvaleri commented on clients/src/main/resources/common/message/AttachMirrorTopicRequest.json:
```diff
@@ -0,0 +1,33 @@
+  "validVersions": "0",
+  "flexibleVersions": "0+",
+  "fields": [
+    { "name": "Topics", "type": "[]TopicState", "versions": "0+", "about": "The name or topic ID of the topic.",
```
```suggestion
    { "name": "Topics", "type": "[]TopicState", "versions": "0+", "about": "The topic state.",
```

fvaleri commented on clients/src/main/resources/common/message/AttachMirrorTopicRequest.json:
```diff
@@ -0,0 +1,33 @@
+        { "name": "Name", "type": "string", "versions": "0+", "mapKey": true, "entityType": "topicName",
+          "about": "The topic name." },
+        { "name": "MirrorName", "type": "string", "versions": "0+", "nullableVersions": "0+",
+          "about": "The cluster link name."}
```
```suggestion
          "about": "The cluster mirror name."}
```

fvaleri commented on clients/src/main/resources/common/message/DeleteMirrorTopicRequest.json:
```diff
@@ -0,0 +1,33 @@
+  "flexibleVersions": "0+",
+  "fields": [
+    { "name": "clusterLink", "type": "string", "versions": "7+", "ignorable": true,
+      "about": "If true, check that the topics can be created as specified, but don't create anything." },
```
```suggestion
      "about": "The cluster mirror name." },
```

fvaleri reviewed (commented):
I agree that having dedicated APIs is better than forcing existing APIs to support more use cases. When the logic behind the attach/add topic is implemented, are we going to revert changes related to `CreateTopics` API? I think we also need `DeleteMirror` API.

showuon committed:
address review comments

showuon commented:
Thanks for the review. I've addressed the comments.

> I agree that having dedicated APIs is better than forcing existing APIs to support more use cases. When the logic behind the attach/add topic is implemented, are we going to revert changes related to `CreateTopics` API? I think we also need `DeleteMirror` API.

1. Yes, we also need `DeleteMirror` API.
2. I personally don't think we need another API for `create mirror topic` because internally, we're still doing the same `create topic` operations, just add 2 fields: `mirrorName`, `topicId`.

fvaleri reviewed (commented):
LGTM, just last update to Admin, KafkaAdminClient, ForwardingAdmin and Mock method names:

- attachMirrorTopic -> addTopicsToMirror
- deleteMirrorTopic -> removeTopicsFromMirror

showuon commented:
>    attachMirrorTopic -> addTopicsToMirror
>    deleteMirrorTopic -> removeTopicsFromMirror

Thanks. I'll fix that in the branch I'm working on.
